website/docs/tutorials/07-testing/03-fuzzing/strategies.md

```diff
@@ -0,0 +1,134 @@
+and also is capable of other internal features, such as simplifying failing examples.
+
+By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt
+parameters.
+
+```cairo title="Example"
+@external
+func setup_integers() {
+    %{
+        given(
+            a = strategy.integers(10, 20),
+            b = strategy.integers(30, 40),
+        )
+    %}
+    return ();
+}
+
+@external
+func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
+    assert_le(a, b);
+    return ();
+}
+```
+
+This document is a guide to what strategies are available for generating examples and how to build
+them.
+
+## Core strategies
+
+All core strategies are contained in the [`strategy`](../02-cheatcodes/strategy.md)
+cheatcode-namespace.
+
+### `strategy.felts`
+
+```python
+def felts() -> Strategy: ...
+```
+
+Explores all possible felt values.
+
+### `strategy.integers`
+
+```python
+def integers(
+    min_value: Optional[int] = None,
+    max_value: Optional[int] = None,
+) -> Strategy: ...
+```
+
+Generates integer values, possibly bounded to provided range.
+
+Assuming real numbers comparison semantics,
+if `min_value` is not `None` then all values will be greater than or equal to `min_value`,
+and if `max_value` is not `None` then all values will be less than or equal to `max_value`.
+When applied to field elements, the unbounded values may rarely overflow.
+
+```python title="Examples"
+strategy.integers(0, 100)
+strategy.integers(max_value=3000)
+strategy.integers(11)
+```
+
+## Adapting strategies
+
+Often it is the case that a strategy does not produce exactly what is desired and a need to further
+adapt the generated values arise.
+Although this could be done in tests directly, this hurts because adaptation may waste fuzzing
+cycles (by repeatedly testing same values after adaptation) and the code has to be copied in every
+test.
+There are also the [`assume`] and [`reject`] cheatcodes, but this should only be used to discard
+only few unwanted inputs.
+
+Protostar provides ways to build strategies by transforming other ones.
+
+### Mapping
+
+```python
+class Strategy:
+    def map(self, mapping_function: Callable[[int], int]) -> Strategy: ...
+```
+
+Applies provided mapping function to all searched inputs values.
+
+```python title="Example"
+strategy.felts().map(lambda x: x // 2)
+```
+
+### Filtering
+
+```python
+class Strategy:
+    def filter(self, filter_function: Callable[[int], bool]) -> Strategy: ...
+```
+
+Allows rejecting examples matching a condition.
+
+The outcome is similar to using the [`assume`] or [`reject`] cheatcodes, but `filter` does not
+require executing tested Cairo function and thus is more performant.
+Try to use `filter` only to avoid unwanted corner cases rather than attempting to cut out a large
+portion of the searched input values.
+
+The `filter` method is not magic and hard to satisfy conditions may cause the fuzzer to fail.
```

Comment:
I have no idea what this sentence means being honest

Reply:
`filter(lambda: 10 < x < 100)` will cause Hypothesis to raise exceptions that strategies are bad

Reply:
```suggestion
Fuzzer draws random data from the original strategy and only afterwards checks if it passes filter conditions.
If too many variables are restricted, fuzzer will reject test execution.
```

Reply:
Discussed on slack